File: PrebidMobile/Rendering/EventHandlers/PrebidMobileMoPubAdapters/Sources/PrebidMoPubNativeAdAdapter.m
//
//  PBMApolloNativeAdAdapter.m
//  OpenXApolloMoPubAdapters
//
//  Copyright © 2021 OpenX. All rights reserved.
//

#import <MoPub.h>

#import <PrebidMobileRendering/PBMNativeAd.h>

#import "PrebidMoPubNativeAdAdapter.h"

@interface PrebidMoPubNativeAdAdapter () <PBMNativeAdUIDelegate, PBMNativeAdTrackingDelegate>

@property (nonatomic, readonly) MediaView *mediaView;

@end

@implementation PrebidMoPubNativeAdAdapter

@synthesize properties = _properties;

- (instancetype)initWithPBMNativeAd:(PBMNativeAd *)nativeAd {
    if (!(self = [super init])) {
        return nil;
    }
    
    _nativeAd = nativeAd;
    _nativeAd.uiDelegate = self;
    _nativeAd.trackingDelegate = self;
    
    NSMutableDictionary *properties = [NSMutableDictionary dictionary];
    
    properties[kAdTitleKey] = nativeAd.title;
    properties[kAdTextKey] = nativeAd.text;
    NSString * const sponsored = [nativeAd dataObjectsOfType:PBMDataAssetType_Sponsored].firstObject.value;
    properties[kAdSponsoredByCompanyKey] = sponsored;
    properties[kAdCTATextKey] = nativeAd.callToAction;
    
    NSString * const iconUrl = nativeAd.iconURL;
    if (iconUrl.length > 0) {
        properties[kAdIconImageKey] = iconUrl;
    }
    
    NSString * const imageUrl = nativeAd.imageURL;
    if (imageUrl.length > 0) {
        properties[kAdMainImageKey] = imageUrl;
    }
    
    NSString * const ratingString = [nativeAd dataObjectsOfType:PBMDataAssetType_Rating].firstObject.value;
    if (ratingString) {
        NSNumberFormatter *formatter = [[NSNumberFormatter alloc] init];
        formatter.numberStyle = NSNumberFormatterDecimalStyle;
        NSNumber *ratingNum = [formatter numberFromString:ratingString];
        properties[kAdStarRatingKey] = ratingNum;
    }
    
    MediaData *mediaData = nativeAd.videoAd.mediaData;
    
    if (mediaData) {
        _mediaView = [[MediaView alloc] init];
        properties[kAdMainMediaViewKey] = _mediaView;
        [_mediaView load:mediaData];
    }
    
    _properties = properties;
    
    return self;
}

#pragma mark - MPNativeAdAdapter

- (NSURL *)defaultActionURL {
    return nil;
}

- (BOOL)enableThirdPartyClickTracking {
    return YES;
}

- (UIView *)mainMediaView {
    return self.mediaView;
}

#pragma mark - PBMNativeAdUIDelegate

- (nullable UIViewController *)viewPresentationControllerForNativeAd:(nonnull PBMNativeAd *)nativeAd {
    return [self.delegate viewControllerForPresentingModalView];
}

- (void)nativeAdWillLeaveApplication:(PBMNativeAd *)nativeAd {
    MPLogEvent([MPLogEvent adWillLeaveApplicationForAdapter:NSStringFromClass(self.class)]);
    [self.delegate nativeAdWillLeaveApplicationFromAdapter:self];
}

- (void)nativeAdWillPresentModal:(PBMNativeAd *)nativeAd {
    MPLogEvent([MPLogEvent adWillPresentModalForAdapter:NSStringFromClass(self.class)]);
    [self.delegate nativeAdWillPresentModalForAdapter:self];
}

- (void)nativeAdDidDismissModal:(PBMNativeAd *)nativeAd {
    MPLogEvent([MPLogEvent adDidDismissModalForAdapter:NSStringFromClass(self.class)]);
    [self.delegate nativeAdDidDismissModalForAdapter:self];
}

#pragma mark - PBMNativeAdTrackingDelegate

- (void)nativeAdDidLogClick:(PBMNativeAd *)nativeAd {
    if ([self.delegate respondsToSelector:@selector(nativeAdDidClick:)]) {
        [self.delegate nativeAdDidClick:self];
        MPLogEvent([MPLogEvent adTappedForAdapter:NSStringFromClass(self.class)]);
    } else {
        MPLogInfo(@"Delegate does not implement click tracking callback. Clicks likely not being tracked.");
    }
}

- (void)nativeAd:(PBMNativeAd *)nativeAd didLogEvent:(PBMNativeEventType)nativeEvent {
    if (nativeEvent == PBMNativeEventType_Impression) {
        if ([self.delegate respondsToSelector:@selector(nativeAdWillLogImpression:)]) {
            MPLogEvent([MPLogEvent adShowSuccessForAdapter:NSStringFromClass(self.class)]);
            MPLogEvent([MPLogEvent adWillAppearForAdapter:NSStringFromClass(self.class)]);
            [self.delegate nativeAdWillLogImpression:self];
        } else {
            MPLogInfo(@"Delegate does not implement impression tracking callback. Impressions likely not being tracked.");
        }
    }
}

@end

File: PrebidMobile/Rendering/EventHandlers/PrebidMobileMoPubAdapters/Sources/PrebidMoPubNativeAdAdapter.swift
/*   Copyright 2018-2021 Prebid.org, Inc.

 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, software
 distributed under the License is distributed on an "AS IS" BASIS,
 WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 See the License for the specific language governing permissions and
 limitations under the License.
 */

import Foundation

import MoPubSDK

//import PrebidMobile
import PrebidMobile

@objc(PrebidMoPubNativeAdAdapter)
class PrebidMoPubNativeAdAdapter:
    NSObject,
    MPNativeAdAdapter,
    NativeAdUIDelegate,
    NativeAdTrackingDelegate
{
    
    // MARK: - Public Properties
    
    weak var delegate: MPNativeAdAdapterDelegate?
    var nativeAd: NativeAd
    
    // MARK: - Internal Properties
    
    var mediaView: MediaView?
    
    // MARK: - Public Methods
    
    init(nativeAd: NativeAd) {
        self.nativeAd = nativeAd
        
        super.init()
        
        self.nativeAd.uiDelegate = self
        self.nativeAd.trackingDelegate = self
        
        properties[kAdTitleKey] = nativeAd.title
        properties[kAdTextKey] = nativeAd.text
       
        let sponsored = nativeAd.dataObjects(of: .sponsored).first?.value
       
        properties[kAdSponsoredByCompanyKey] = sponsored;
        properties[kAdCTATextKey] = nativeAd.callToAction;
        
        if !nativeAd.iconURL.isEmpty {
            properties[kAdIconImageKey] = nativeAd.iconURL.isEmpty;
        }
        
        if !nativeAd.imageURL.isEmpty {
            properties[kAdMainImageKey] = nativeAd.imageURL;

        }
        
        if let ratingString = nativeAd.dataObjects(of: .rating).first?.value {
            let formatter = NumberFormatter()
            formatter.numberStyle = .decimal
            let ratingNum = formatter.number(from: ratingString)
            properties[kAdStarRatingKey] = ratingNum
        }
        
        if let mediaData = nativeAd.videoAd?.mediaData {
            mediaView = MediaView()
            properties[kAdMainMediaViewKey] = mediaView
            mediaView?.load(mediaData)
        }
    }
    
    // MARK: - MPNativeAdAdapter
    
    var properties = [AnyHashable : Any]()
    
    var defaultActionURL: URL? {
        nil
    }
    
    func enableThirdPartyClickTracking() -> Bool {
        true
    }
    
    func mainMediaView() -> UIView? {
        self.mediaView
    }
    
    // MARK: - NativeAdUIDelegate
    
    func viewPresentationControllerForNativeAd(_ nativeAd: NativeAd) -> UIViewController? {
        delegate?.viewControllerForPresentingModalView()
    }
    
    func nativeAdWillLeaveApplication(_ nativeAd: NativeAd) {
        MPLogging.logEvent(MPLogEvent.adWillLeaveApplication(forAdapter: Self.className()), source: nil, from: nil)
        delegate?.nativeAdWillLeaveApplication(from: self)
    }
    
    func nativeAdWillPresentModal(_ nativeAd: NativeAd) {
        MPLogging.logEvent(MPLogEvent.adWillPresentModal(forAdapter: Self.className()), source: nil, from: nil)
        delegate?.nativeAdWillPresentModal(for: self)
    }
    
    func nativeAdDidDismissModal(_ nativeAd: NativeAd) {
        MPLogging.logEvent(MPLogEvent.adDidDismissModal(forAdapter: Self.className()), source: nil, from: nil)
        delegate?.nativeAdDidDismissModal(for: self)
    }
    
    // MARK: - NativeAdTrackingDelegate
    
    func nativeAdDidLogClick(_ nativeAd: NativeAd) {
        guard let delegate = delegate,
           delegate.responds(to: #selector(MPNativeAdAdapterDelegate.nativeAdDidClick)) else {
            return
        }
        
        MPLogging.logEvent(MPLogEvent.adTapped(forAdapter: Self.className()), source: nil, from: nil)
        delegate.nativeAdDidClick?(self)
    }
    
    func nativeAd(_ nativeAd: NativeAd, didLogEvent nativeEvent: NativeEventType) {
        guard nativeEvent == .impression,
              let delegate = delegate,
              delegate.responds(to: #selector(MPNativeAdAdapterDelegate.nativeAdWillLogImpression)) else {
            return
        }
        
        MPLogging.logEvent(MPLogEvent.adShowSuccess(forAdapter: Self.className()), source: nil, from: nil)
        MPLogging.logEvent(MPLogEvent.adWillAppear(forAdapter: Self.className()), source: nil, from: nil)
        
        delegate.nativeAdWillLogImpression?(self)
    }
}
